### PR TITLE
Simulated the synchrotron topup PVs and had them "link" #22

### PR DESCRIPTION
**Pull request by evalott100**
Simulated the current loss and current gain of the synchrotron storage ring. The current loss and gain increment are calculated so they will take 10 minutes and 15 seconds respectively. The countdown is also simulated: based on the change of the current (this will only really work as a method if the rate of change is constant). Also wrote tests.

### CODE REVIEW

**Review by abbiemery — CHANGES_REQUESTED** [review]
Looking good, just a few suggestions with a refactor on the synchrotron_current. 

Also any chance you could change the `s03_config` one as well to match.

File: tickit_devices/synchrotron/synchrotron_current.py [hunk 1]
```diff
@@ -113,7 +150,10 @@ def on_db_load(self) -> None:
 class SynchrotronCurrent(ComponentConfig):
     """Synchrotron current component."""
 
-    initial_current: Optional[float] = None
+    initial_current: Optional[float]
+    charge_loss_dist = {"mu": -2.0, "sigma": 1, "scale": 0.0333}
+    charge_gain_dist = {"mu": 5.0, "sigma": 1.15, "scale": 0.1}
```
Comment (DominicOram) [inline tickit_devices/synchrotron/synchrotron_current.py:155]: Should: These aren't used anymore, can you remove them?

File: tickit_devices/synchrotron/synchrotron_topup.py [hunk 2]
```diff
@@ -59,12 +75,39 @@ def update(self, time: SimTime, inputs: Inputs) -> DeviceUpdate[Outputs]:
...
+
+        # calculate the number of updates to go until topup stage begins or
+        # ends: based upon the change of the current.
+        updates_to_go = abs(
+            topup_fill * (300 - new_current) / current_difference
+            + (not topup_fill) * (new_current - 270) / current_difference
+        )
+
+        if self.last_time:
+            self.period = (time - self.last_time) * 1e-9
```
Comment (DominicOram) [inline tickit_devices/synchrotron/synchrotron_topup.py:92]: Suggestion: I think it makes more sense to have period as just a local variable and set it here in both cases of the `last_time` being None or not

File: tickit_devices/synchrotron/synchrotron_current.py [hunk 3]
```diff
@@ -51,8 +72,24 @@ def update(self, time: SimTime, inputs: Inputs) -> DeviceUpdate[Outputs]:
...
                 The produced update event which contains the value of the beam current.
         """
+        # check to see if topup fill should be activated/deactivated
+        if self.topup_fill:
+            self.topup_fill = self.beam_current < self.target_current
+        else:
+            self.topup_fill = self.beam_current <= 270
+
+        if self.last_update_time:
+            self.period = time - self.last_update_time
```
Comment (DominicOram) [inline tickit_devices/synchrotron/synchrotron_current.py:82]: Suggestion: As below, I think it's nicer to heave period just set in here.

File: tickit_devices/synchrotron/synchrotron_current.py [hunk 4]
```diff
@@ -51,8 +72,24 @@ def update(self, time: SimTime, inputs: Inputs) -> DeviceUpdate[Outputs]:
             DeviceUpdate[Outputs]:
                 The produced update event which contains the value of the beam current.
         """
+        # check to see if topup fill should be activated/deactivated
+        if self.topup_fill:
+            self.topup_fill = self.beam_current < self.target_current
```
Comment (DominicOram) [inline tickit_devices/synchrotron/synchrotron_current.py:77]: Should: Could we have a test that checks the topup fill stops when expected

**Review by DominicOram — CHANGES_REQUESTED** [review]
Generally looks good, some minor comments. Also, there are a number of hardcoded assumptions about the behavior i.e.
* What current to topup at
* The loss increment (or time taken between topups)
* The gain increment (or time taken to topup)

Can you pull them out into either constants or (ideally) into variables that we can pass into the config?

File: tickit_devices/synchrotron/synchrotron_topup.py (hunk 2 above)
Comment (evalott100) [reply tickit_devices/synchrotron/synchrotron_topup.py:92]: Good idea

**Pull request merged**